tests/bench_ztimer/main.c

```diff
@@ -0,0 +1,296 @@
+ */
+static unsigned _triggers;
+
+/*
+ * The test assumes that first, middle and last will always end up in at the
+ * same index within the timer queue.  In order to compensate for the time that
+ * previous operations take themselves, the interval is corrected. The
+ * variables "start" and "_base" are used for that.
+ */
+uint32_t _base;
+
+static void _callback(void *arg) {
+    unsigned *triggers = arg;
+    *triggers += 1;
+}
+
+/* returns the interval for timer 'n' that has to be set in order to insert it
+ * into position n */
+static uint32_t _timer_val(unsigned n)
+{
+    return _base + (SPREAD * n);
+}
+
+/* set timer 'n' to its intended position 'n' */
+static void _timer_set(unsigned n)
+{
+    ztimer_set(ZTIMER, &_timers[n], _timer_val(n));
+}
+
+/* remove timer 'n' */
+static void _timer_remove(unsigned n)
+{
+    ztimer_remove(ZTIMER, &_timers[n]);
+}
+
+static void _print_result(const char *desc, unsigned n, uint32_t total)
+{
+    printf("%30s %8"PRIu32" / %u = %"PRIu32"\n", desc, total, n, total/n);
+}
+
+int main(void)
+{
+    puts("ztimer benchmark application.\n");
+
+    unsigned n;
+    uint32_t before, diff, start;
+
+    /* initializing timer structs */
+    for (unsigned int n = 0; n < NUMOF_TIMERS; n++) {
+        _timers[n].callback = _callback;
+        _timers[n].arg = &_triggers;
+    }
+
+    start = ztimer_now(ZTIMER_USEC);
+
+    /*
+     * test setting one set timer REPEAT times
+     *
+     */
+    _base = BASE;
+    before = ztimer_now(ZTIMER_USEC);
+    for (n = 0; n < REPEAT; n++) {
+        _timer_set(0);
+    }
+
+    diff = ztimer_now(ZTIMER_USEC) - before;
+
+    _print_result("set() one", REPEAT, diff);
+    expect(!_triggers);
+
+    /*
+     * test removing one unset timer REPEAT times
+     *
+     */
+    before = ztimer_now(ZTIMER_USEC);
+    for (n = 0; n < REPEAT; n++) {
+        _timer_remove(0);
+    }
+
+    diff = ztimer_now(ZTIMER_USEC) - before;
+
+    _print_result("remove() one", REPEAT, diff);
+    expect(!_triggers);
+
+    /*
+     * test setting / removing one timer REPEAT times
+     *
+     */
+    before = ztimer_now(ZTIMER_USEC);
+    _base = BASE  - (before - start);
+    for (n = 0; n < REPEAT; n++) {
+        _timer_set(0);
+        _timer_remove(0);
+    }
+
+    diff = ztimer_now(ZTIMER_USEC) - before;
+
+    _print_result("set() + remove() one", REPEAT, diff);
+    expect(!_triggers);
+
+    /*
+     * test setting NUMOF_TIMERS timers with increasing targets
+     *
+     */
+    before = ztimer_now(ZTIMER_USEC);
+    _base = BASE  - (before - start);
+    for (unsigned int n = 0; n < NUMOF_TIMERS; n++) {
+        _timer_set(n);
+    }
+
+    diff = ztimer_now(ZTIMER_USEC) - before;
+
+    _print_result("set() many increasing target", NUMOF_TIMERS, diff);
+    expect(!_triggers);
+
+    /*
+     * test re-setting first timer REPEAT times
+     *
+     */
+    before = ztimer_now(ZTIMER_USEC);
+    _base = BASE  - (before - start);
```

Comment:
What's the idea behind this?? Its dynamically changing the offest as the test progresses, this makes some timers trigger unexpectedly depending on how long the test took

Reply:
BASE is supposed to be set way in the future, so that during the test, no timer actually triggers.
That's verified by `expect(!triggers)` throughout the test.